File: network/pi_prime_network.cpp
#include <iostream>
#include <string>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>

class PiPrimeNetwork {
public:
    PiPrimeNetwork(int port) {
        socket_ = socket(AF_INET, SOCK_STREAM, 0);
        struct sockaddr_in addr;
        addr.sin_family = AF_INET;
        addr.sin_port = htons(port);
        inet_pton(AF_INET, "127.0.0.1", &addr.sin_addr);
        bind(socket_, (struct sockaddr*)&addr, sizeof(addr));
        listen(socket_, 5);
    }

    void listen() {
        while (true) {
            struct sockaddr_in client_addr;
            socklen_t client_len = sizeof(client_addr);
            int client_socket = accept(socket_, (struct sockaddr*)&client_addr, &client_len);
            handleConnection(client_socket);
        }
    }

    void handleConnection(int client_socket) {
        char input_bytes[16];
        read(client_socket, input_bytes, 16);
        long long input = *((long long*)input_bytes);
        bool output = isPrime(input);
        char output_bytes[1] = {(char)(output ? 1 : 0)};
        write(client_socket, output_bytes, 1);
    }

    bool isPrime(long long n) {
        if (n < 2) return false;
        for (long long i = 2; i * i <= n; i++) {
            if (n % i == 0) return false;
        }
        return true;
    }

private:
    int socket_;
};

int main() {
    PiPrimeNetwork network(8080);
    network.listen();
    return 0;
}

File: network/pi_prime_network.py
import socket
import struct

class PiPrimeNetwork:
    def __init__(self, port):
        self.listener = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.listener.bind(("localhost", port))
        self.listener.listen(5)

    def listen(self):
        while True:
            conn, addr = self.listener.accept()
            self.handle_connection(conn)

    def handle_connection(self, conn):
        input_bytes = conn.recv(16)
        input_int = struct.unpack("<Q", input_bytes)[0]
        output_bool = self.is_prime(input_int)
        output_bytes = struct.pack("<?", output_bool)
        conn.sendall(output_bytes)

    def is_prime(self, n):
        if n < 2:
            return False
        for i in range(2, int(n ** 0.5) + 1):
            if n % i == 0:
                return False
        return True

if __name__ == "__main__":
    network = PiPrimeNetwork(8080)
    network.listen()
